### PR TITLE
Output UnionType str rep with () instead of []

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69502

Differential Revision: [D32902781](https://our.internmc.facebook.com/intern/diff/D32902781)

### CODE REVIEW

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32902781).

File: test/test_jit.py [hunk 1]
```diff
@@ -3635,6 +3635,19 @@ def check_subclass_warn(input: torch.LongTensor) -> torch.LongTensor:
...
+        def test_func(a: Union[int, float], b: Optional[int]):
+            return 0
+
+        scripted_func = torch.jit.script(test_func)
+        graph_rep = str(scripted_func.graph)
+        code_rep = str(scripted_func.code)
+        # TS graph IR for Union should be annotated as Union()
+        FileCheck().check("Union(").check("int?").run(graph_rep)
+        # Serialized code for Union should be annotated as Union[]
+        FileCheck().check("Union[").check("Optional[int]").run(code_rep)
```
Comment (gmagogsfm) [inline test/test_jit.py:3648]: Indentation is weird?

File: test/test_jit.py [hunk 2]
```diff
@@ -3635,6 +3635,19 @@ def check_subclass_warn(input: torch.LongTensor) -> torch.LongTensor:
...
+        self.checkScript(test_func, (5, 6))
+
```
Comment (gmagogsfm) [inline test/test_jit.py:3650]: Maybe call `parse_ir` on `graph_rep` to verify round-trip-ability.

File: aten/src/ATen/core/type.cpp [hunk 3]
```diff
@@ -1141,16 +1144,16 @@ std::string UnionType::unionStr(TypePrinter printer, bool is_annotation_str) con
...
     }
   }
-  ss << "]";
+  ss << close_delimeter;
   return ss.str();
 }
 
 std::string UnionType::str() const {
-  return this->unionStr(nullptr, /*is_annotation_str=*/false);
+  return this->unionStr(nullptr, /*is_annotation_str=*/false, "(", ")");
```
Comment (gmagogsfm) [inline aten/src/ATen/core/type.cpp:1152]: You can probably use `is_annotation_str` to branch on which delimiter to use.

**Review by gmagogsfm — CHANGES_REQUESTED** [review]

File: test/test_jit.py (hunk 1 above)
Comment (tugsbayasgalan) [reply test/test_jit.py]: looks fine to me?

File: test/test_jit.py (hunk 2 above)
Comment (tugsbayasgalan) [reply test/test_jit.py]: sounds good

**Comment by tugsbayasgalan** [comment]
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32902781).

File: aten/src/ATen/core/type.cpp (hunk 3 above)
Comment (tugsbayasgalan) [reply aten/src/ATen/core/type.cpp]: sounds good

File: aten/src/ATen/core/type.cpp [hunk 4]
```diff
@@ -1116,7 +1116,10 @@ std::string UnionType::unionStr(TypePrinter printer, bool is_annotation_str) con
     return false;
   };
 
-  ss << "Union[";
+  std::string open_delimeter = is_annotation_str ? "[" : "(";
+  std::string close_delimeter = is_annotation_str ? "]" : ")";
+
+  ss << "Union" + open_delimeter;
```
Comment (gmagogsfm) [inline aten/src/ATen/core/type.cpp:1122]:
```suggestion
  ss << "Union" << open_delimeter;
```

**Review by gmagogsfm — APPROVED** [review]

**Pull request closed without merging**